### DISCUSSION TITLE
In skywalking-java-agent, can I decide whether to report this Segment or not based on the state of the Segment?

### DISCUSSION THREAD

**Discussion by OrezzerO** [apache/skywalking#12904] [Q&A] [answered]
Currently, the sampling strategy of skywalking-java is that whether to sample has already been determined when creating the TracerContext. Is it considered to implement a layer of sampling before reporting to the server? In this way, it will be possible to decide whether to report according to the state of the Segment, such as the time consumption and whether an error has occurred or not.

Are there any implicit constraints that prevent us from doing so?

**Comment by wu-sheng** [comment]
What is the benefit of this? Collecting span and context propagation cost most CPU and memory, you chose to finish that but skip the report, why?

**Reply by OrezzerO** [reply]
Our server is provided by the cloud service provider, and they don't offer the sampling function on the server side. The goals I want to achieve are as follows: 
1. Reduce the storage on the server side; 
3. Don't miss the Segments that contain errors or slow queries.

**Reply by wu-sheng** [reply]
Then,it's not skywalking use case. Sorry, we could not support for other vendors.

**Reply by wu-sheng** [reply]
They cut SkyWalking supported feature, it is their responsibility now.

**Reply by OrezzerO** [reply]
I can understand you, but it's uncertain to push the cloud service provider to make changes. I'd like to modify SkyWalking-Java on my own. What I want to ask the community for help is that I want to know if there are any unfeasible aspects in this matter, or get some suggestions so that I can solve this problem in a better way.

**Reply by wu-sheng** [reply]
> What is the benefit of this? Collecting span and context propagation cost most CPU and memory, you chose to finish that but skip the report, why?

This is what I mean. I can't see the value to do that.
I can't tell why cloud vendor doesn't support that, most likely it is from billing plan perspective(charged by store cost, rather than received data)

**Reply by wu-sheng** [reply, marked as answer]
> that I want to know if there are any unfeasible aspects in this matter

From upstream perspective, it is a low value and high cost sampling mechanism. It should not be done.

**Reply by OrezzerO** [reply]
Got it. Thank you for your reply.
